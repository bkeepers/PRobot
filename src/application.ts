import { WebhookEvent } from '@octokit/webhooks'
import deprecated from 'deprecated-decorator'
import express from 'express'
import { EventEmitter } from 'promise-events'
import { ApplicationFunction } from '.'
import { Context } from './context'
import { GitHubAPI } from './github'
import { GitHubApp } from './github-app'
import { logger } from './logger'
import { LoggerWithTarget, wrapLogger } from './wrap-logger'

/**
 * The `app` parameter available to apps
 *
 * @property {logger} log - A logger
 */
export class Application {
  public events: EventEmitter
  public router: express.Router
  public catchErrors: boolean
  public log: LoggerWithTarget

  private github: GitHubApp

  constructor (options?: Options) {
    const opts = options || {} as any
    this.events = new EventEmitter()
    this.log = wrapLogger(logger, logger)
    this.github = opts.github
    this.catchErrors = opts.catchErrors || false
    this.router = opts.router || express.Router() // you can do this?
  }

  /**
   * Loads a Probot plugin
   * @param plugin - Probot plugin to load
   */
  public load (app: ApplicationFunction | ApplicationFunction[]): Application {
    if (Array.isArray(app)) {
      app.forEach(a => this.load(a))
    } else {
      app(this)
    }

    return this
  }

  public async receive (event: WebhookEvent) {
    if ((event as any).event) {
      // tslint:disable-next-line:no-console
      console.warn(new Error('Propery `event` is deprecated, use `name`'))
      event = { name: (event as any).event, ...event }
    }

    return Promise.all([
      this.events.emit('*', event),
      this.events.emit(event.name, event),
      this.events.emit(`${ event.name }.${ event.payload.action }`, event)
    ])
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = app => {
   *   // Get an express router to expose new HTTP endpoints
   *   const route = app.route('/my-app');
   *
   *   // Use any middleware
   *   route.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   route.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes
   * @returns an [express.Router](http://expressjs.com/en/4x/api.html#router)
   */
  public route (path?: string): express.Router {
    if (path) {
      const router = express.Router()
      this.router.use(path, router)
      return router
    } else {
      return this.router
    }
  }

  /**
   * Listen for [GitHub webhooks](https://developer.github.com/webhooks/),
   * which are fired for almost every significant action that users take on
   * GitHub.
   *
   * @param event - the name of the [GitHub webhook
   * event](https://developer.github.com/webhooks/#events). Most events also
   * include an "action". For example, the * [`issues`](
   * https://developer.github.com/v3/activity/events/types/#issuesevent)
   * event has actions of `assigned`, `unassigned`, `labeled`, `unlabeled`,
   * `opened`, `edited`, `milestoned`, `demilestoned`, `closed`, and `reopened`.
   * Often, your bot will only care about one type of action, so you can append
   * it to the event name with a `.`, like `issues.closed`.
   *
   * ```js
   * app.on('push', context => {
   *   // Code was just pushed.
   * });
   *
   * app.on('issues.opened', context => {
   *   // An issue was just opened.
   * });
   * ```
   *
   * @param callback - a function to call when the
   * webhook is received.
   */
  public on (eventName: string | string[], callback: (context: Context) => Promise<void>) {
    if (typeof eventName === 'string') {
      return this.events.on(eventName, async (event: WebhookEvent) => {
        try {
          await callback(await this.github.createContext(event))
        } catch (err) {
          this.log.error({ err, event, id: event.id })
          if (!this.catchErrors) {
            throw err
          }
        }
      })
    } else {
      eventName.forEach(e => this.on(e, callback))
    }
  }

  @deprecated('github.jwt')
  public app (): string {
    return this.github.jwt()
  }

  @deprecated('github.auth')
  public auth (id?: number, log = this.log): Promise<GitHubAPI> {
    return this.github.auth(id, log)
  }
}

export interface Options {
  github: GitHubApp
  router?: express.Router
  catchErrors?: boolean
}
