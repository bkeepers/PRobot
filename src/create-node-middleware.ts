import { RequestListener, IncomingMessage, ServerResponse } from "http";
import { NextFunction } from "express";

import { ApplicationFunction } from "./types";
import { MiddlewareOptions } from "./types";

export function createNodeMiddleware(
  appFn: ApplicationFunction,
  { probot }: MiddlewareOptions
): RequestListener {
  return (
    request: IncomingMessage,
    response: ServerResponse,
    next?: NextFunction
  ) => {
    probot.load(appFn);
    probot.webhooks.middleware(request, response, next);
  };
}
