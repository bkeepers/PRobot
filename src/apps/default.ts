import path from "path";
import express from "express";
import { Application } from "../application";

export = ({
  app,
  getRouter,
}: {
  app: Application;
  getRouter: () => express.Router;
}) => {
  const router = getRouter();

  router.get("/probot", (req, res) => {
    let pkg;
    try {
      pkg = require(path.join(process.cwd(), "package.json"));
    } catch (e) {
      pkg = {};
    }

    res.render("probot.hbs", pkg);
  });
  router.get("/", (req, res, next) => res.redirect("/probot"));
};
