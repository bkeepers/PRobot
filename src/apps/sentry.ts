import sentryStream from 'bunyan-sentry-stream'
import Raven from 'raven'
import { Application } from '../application'

export = (app: Application) => {
  // If sentry is configured, report all logged errors
  if (process.env.SENTRY_DSN) {
    app.log.debug(process.env.SENTRY_DSN, 'Errors will be reported to Sentry')
    Raven.disableConsoleAlerts()
    Raven.config(process.env.SENTRY_DSN, {
      captureUnhandledRejections: true,
      tags: {
        version: process.env.HEROKU_RELEASE_VERSION as string
      },
      release: process.env.HEROKU_SLUG_COMMIT,
      environment: process.env.NODE_ENV || 'development',
      autoBreadcrumbs: true
    }).install()

    app.log.target.addStream(sentryStream(Raven))
  }
}
