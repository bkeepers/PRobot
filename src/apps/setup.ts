import bodyParser from "body-parser";
import { exec } from "child_process";
import { Request, Response } from "express";
import updateDotenv from "update-dotenv";
import { Application } from "../application";
import { ManifestCreation } from "../manifest-creation";

import { getLoggingMiddleware } from "../server/logging-middleware";

// TODO: after #1335 is merged, pass 'host' as well
export const setupAppFactory = (port: number | undefined) =>
  async function setupApp(app: Application) {
    const setup: ManifestCreation = new ManifestCreation();

    // If not on Glitch or Production, create a smee URL
    if (
      process.env.NODE_ENV !== "production" &&
      !(process.env.PROJECT_DOMAIN || process.env.WEBHOOK_PROXY_URL)
    ) {
      await setup.createWebhookChannel();
    }

    const route = app.route();

    route.use(getLoggingMiddleware(app.log));

    printWelcomeMessage(app, port);

    route.get("/probot", async (req, res) => {
      const baseUrl = getBaseUrl(req);
      const pkg = setup.pkg;
      const manifest = setup.getManifest(pkg, baseUrl);
      const createAppUrl = setup.createAppUrl;
      // Pass the manifest to be POST'd
      res.render("setup.hbs", { pkg, createAppUrl, manifest });
    });

    route.get("/probot/setup", async (req: Request, res: Response) => {
      const { code } = req.query;
      const response = await setup.createAppFromCode(code);

      restartGlitchApp(app);

      res.redirect(`${response}/installations/new`);
    });

    route.get("/probot/import", async (_req, res) => {
      const { WEBHOOK_PROXY_URL } = process.env;
      res.render("import.hbs", { WEBHOOK_PROXY_URL });
    });

    route.post("/probot/import", bodyParser.json(), async (req, res) => {
      const { appId, pem, webhook_secret } = req.body;
      if (!appId || !pem || !webhook_secret) {
        res.status(400).send("appId and/or pem and/or webhook_secret missing");
        return;
      }
      updateDotenv({
        APP_ID: appId,
        PRIVATE_KEY: `"${pem}"`,
        WEBHOOK_SECRET: webhook_secret,
      });
      res.end();
      restartGlitchApp(app);
    });

    route.get("/probot/success", async (req, res) => {
      res.render("success.hbs");
    });

    route.get("/", (req, res, next) => res.redirect("/probot"));
  };

function printWelcomeMessage(app: Application, port: number | undefined) {
  // use glitch env to get correct domain welcome message
  // https://glitch.com/help/project/
  const domain =
    process.env.PROJECT_DOMAIN || `http://localhost:${port || 3000}`;

  [
    ``,
    `Welcome to Probot!`,
    `Probot is in setup mode, webhooks cannot be received and`,
    `custom routes will not work until APP_ID and PRIVATE_KEY`,
    `are configured in .env.`,
    `Please follow the instructions at ${domain} to configure .env.`,
    `Once you are done, restart the server.`,
    ``,
  ].forEach((line) => {
    app.log.info(line);
  });
}

function getBaseUrl(req: Request): string {
  const protocols = req.headers["x-forwarded-proto"] || req.protocol;
  const protocol =
    typeof protocols === "string" ? protocols.split(",")[0] : protocols[0];
  const host = req.headers["x-forwarded-host"] || req.get("host");
  const baseUrl = `${protocol}://${host}`;
  return baseUrl;
}

function restartGlitchApp(app: Application) {
  // If using glitch, restart the app
  if (process.env.PROJECT_DOMAIN) {
    exec("refresh", (error) => {
      if (error) {
        app.log.error(error);
      }
    });
  }
}
