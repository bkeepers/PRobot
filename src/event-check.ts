import { Application } from './application'
import { GitHubAPI } from './github'

let appMetadata: ReturnType<GitHubAPI['apps']['getAuthenticated']> | null = null

// To avoid displaying a message multiple times, we keep track of which messages
// have already been displayed.
const hasDisplayedWarning = {
  failedRetrievingMeta: false
}

async function eventCheck (app: Application, eventName: string) {
  if (isEventCheckEnabled() === false) {
    return
  }

  const baseEventName = eventName.split('.')[0]
  if (!(await isSubscribedToEvent(app, baseEventName)) && !hasDisplayedWarning.failedRetrievingMeta) {
    app.log.error(`Your app is attempting to listen to the "${eventName}" event, but your GitHub App is not subscribed to the "${baseEventName}" event.`)
  }
}

function isEventCheckEnabled () {
  if (process.env.DISABLE_EVENT_CHECK && process.env.DISABLE_EVENT_CHECK.toLowerCase() === 'true') {
    return false
  }

  if (process.env.NODE_ENV && process.env.NODE_ENV.toLowerCase() !== 'development') {
    return false
  }

  return true
}

/**
 * @param {Application} app
 * @param {string} baseEventName The base part of an event name refers to the
 * text of an event name before the first period mark (e.g. the `issues` part in
 * `issues.opened`).
 * @returns Returns `false` when it is known the application is not subscribed
 *  to the `baseEventName` event. Returns `true` in all other instances.
 *
 *  **Return Caveat Notice:** This function will return `true` if event-check
 *  fails to retrieve subscribed event data. For that reason, it is recommended
 *  to only treat `false` return values as accurate.
 */
async function isSubscribedToEvent (app: Application, baseEventName: string) {
  let events
  if (baseEventName === '*') {
    return true
  }

  try {
    events = (await retrieveAppMeta(app)).data.events
  } catch (e) {
    if (!hasDisplayedWarning.failedRetrievingMeta) {
      app.log.warn(e)
    }
    hasDisplayedWarning.failedRetrievingMeta = true
    return false
  }

  return events.includes(baseEventName)
}

async function retrieveAppMeta (app: Application) {
  if (appMetadata) return appMetadata

  appMetadata = new Promise(async (resolve, reject) => {
    const api = await app.auth()
    try {
      const meta = await api.apps.getAuthenticated()
      return resolve(meta)
    } catch (e) {
      app.log.trace(e)
      /**
       * There are a few reasons why Probot might be unable to retrieve
       * application metadata.
       *
       * - Probot may not be connected to the Internet.
       * - The GitHub API is not responding to requests (see
       *   https://www.githubstatus.com/).
       * - The user has incorrectly configured environment variables (e.g.
       *   APP_ID, PRIVATE_KEY, etc.) used for authentication between the Probot
       *   app and the GitHub API.
       */
      return reject('Probot is unable to retrieve app information from GitHub for event subscription verification.')
    }
  })

  return appMetadata
}

export default eventCheck
