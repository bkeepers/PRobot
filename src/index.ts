// tslint:disable-next-line: no-var-requires
require("dotenv").config();

import { Server } from "http";

import express from "express";
import Redis from "ioredis";
import LRUCache from "lru-cache";
import { Deprecation } from "deprecation";
import pinoHttp from "pino-http";
import { getPrivateKey } from "@probot/get-private-key";

import type { WebhookEvent, Webhooks } from "@octokit/webhooks";
import type { Logger } from "pino";

import { Application } from "./application";
import { setupAppFactory } from "./apps/setup";
import { Context, WebhookPayloadWithRepository } from "./context";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { getLog } from "./helpers/get-log";
import { resolveAppFunction } from "./helpers/resolve-app-function";
import { createServer } from "./server/create-server";
import { createWebhookProxy } from "./helpers/webhook-proxy";
import { getErrorHandler } from "./helpers/get-error-handler";
import {
  ApplicationFunction,
  DeprecatedLogger,
  ProbotWebhooks,
  State,
} from "./types";
import { getProbotOctokitWithDefaults } from "./octokit/get-probot-octokit-with-defaults";
import { aliasLog } from "./helpers/alias-log";
import { logWarningsForObsoleteEnvironmentVariables } from "./helpers/log-warnings-for-obsolete-environment-variables";
import { getWebhooks } from "./octokit/get-webhooks";
import { webhookEventCheck } from "./helpers/webhook-event-check";
import { auth } from "./auth";
import { load } from "./load";
import { getRouter } from "./get-router";

logWarningsForObsoleteEnvironmentVariables();

export interface Options {
  // same options as Application class
  privateKey?: string;
  githubToken?: string;
  id?: number;
  Octokit?: typeof ProbotOctokit;
  log?: Logger;
  redisConfig?: Redis.RedisOptions;
  secret?: string;
  webhookPath?: string;

  // Probot class-specific options
  /**
   * @deprecated `cert` options is deprecated. Use `privateKey` instead
   */
  cert?: string;
  port?: number;
  host?: string;
  webhookProxy?: string;
  /**
   * @deprecated set `Octokit` to `ProbotOctokit.defaults({ throttle })` instead
   */
  throttleOptions?: any;
}

// tslint:disable:no-var-requires
const defaultAppFns: ApplicationFunction[] = [require("./apps/default")];
// tslint:enable:no-var-requires

export class Probot {
  public static async run(appFn: ApplicationFunction | string[]) {
    const pkgConf = require("pkg-conf");
    const program = require("commander");

    const readOptions = (): Options => {
      if (Array.isArray(appFn)) {
        program
          .usage("[options] <apps...>")
          .option(
            "-p, --port <n>",
            "Port to start the server on",
            process.env.PORT || 3000
          )
          .option(
            "-H --host <host>",
            "Host to start the server on",
            process.env.HOST
          )
          .option(
            "-W, --webhook-proxy <url>",
            "URL of the webhook proxy service.`",
            process.env.WEBHOOK_PROXY_URL
          )
          .option(
            "-w, --webhook-path <path>",
            "URL path which receives webhooks. Ex: `/webhook`",
            process.env.WEBHOOK_PATH
          )
          .option("-a, --app <id>", "ID of the GitHub App", process.env.APP_ID)
          .option(
            "-s, --secret <secret>",
            "Webhook secret of the GitHub App",
            process.env.WEBHOOK_SECRET
          )
          .option(
            "-P, --private-key <file>",
            "Path to certificate of the GitHub App",
            process.env.PRIVATE_KEY_PATH
          )
          .parse(appFn);

        return {
          privateKey:
            getPrivateKey({ filepath: program.privateKey }) || undefined,
          id: program.app,
          port: program.port,
          host: program.host,
          secret: program.secret,
          webhookPath: program.webhookPath,
          webhookProxy: program.webhookProxy,
        };
      }
      const privateKey = getPrivateKey();
      return {
        privateKey: (privateKey && privateKey.toString()) || undefined,
        id: Number(process.env.APP_ID),
        port: Number(process.env.PORT) || 3000,
        host: process.env.HOST,
        secret: process.env.WEBHOOK_SECRET,
        webhookPath: process.env.WEBHOOK_PATH,
        webhookProxy: process.env.WEBHOOK_PROXY_URL,
      };
    };

    const options = readOptions();
    const probot = new Probot(options);

    if (!options.id || !options.privateKey) {
      if (process.env.NODE_ENV === "production") {
        if (!options.id) {
          throw new Error(
            "Application ID is missing, and is required to run in production mode. " +
              "To resolve, ensure the APP_ID environment variable is set."
          );
        } else if (!options.privateKey) {
          throw new Error(
            "Certificate is missing, and is required to run in production mode. " +
              "To resolve, ensure either the PRIVATE_KEY or PRIVATE_KEY_PATH environment variable is set and contains a valid certificate"
          );
        }
      }
      probot.load(setupAppFactory(probot.options.host, probot.options.port));
    } else if (Array.isArray(appFn)) {
      const pkg = await pkgConf("probot");
      probot.setup(program.args.concat(pkg.apps || []));
    } else {
      probot.load(appFn);
    }
    probot.start();

    return probot;
  }

  public server: express.Application;
  public webhooks: ProbotWebhooks;
  public log: DeprecatedLogger;
  public version: String;
  public on: ProbotWebhooks["on"];
  public auth: (
    installationId?: number,
    log?: Logger
  ) => Promise<InstanceType<typeof ProbotOctokit>>;

  // These need to be public for the tests to work.
  public options: Options;
  public throttleOptions: any;

  private httpServer?: Server;
  private state: State;

  /**
   * @deprecated this.internalRouter can be removed once we remove the Application class
   */
  private internalRouter: express.Router;

  /**
   * @deprecated use probot.log instead
   */
  public get logger() {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.logger" is deprecated. Use "probot.log" instead`
      )
    );
    return this.log;
  }

  /**
   * @deprecated "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"
   */
  public get router() {
    this.log.warn(
      new Deprecation(
        `[probot] "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return this.internalRouter;
  }

  constructor(options: Options) {
    if (process.env.GHE_HOST && /^https?:\/\//.test(process.env.GHE_HOST)) {
      throw new Error(
        "Your `GHE_HOST` environment variable should not begin with https:// or http://"
      );
    }

    //
    // Probot class-specific options (Express server & Webhooks)
    //
    options.webhookPath = options.webhookPath || "/";
    options.secret = options.secret || "development";

    this.log = aliasLog(options.log || getLog());

    if (options.cert) {
      this.log.warn(
        new Deprecation(
          `[probot] "cert" option is deprecated. Use "privateKey" instead`
        )
      );
      options.privateKey = options.cert;
    }

    if (process.env.INSTALLATION_TOKEN_TTL) {
      this.log.warn(
        '[probot] "INSTALLATION_TOKEN_TTL" environment variable is no longer used. Tokens are renewed as needed at the time of the request now.'
      );
    }

    // TODO: support redis backend for access token cache if `options.redisConfig || process.env.REDIS_URL`
    const cache = new LRUCache<number, string>({
      // cache max. 15000 tokens, that will use less than 10mb memory
      max: 15000,
      // Cache for 1 minute less than GitHub expiry
      maxAge: 1000 * 60 * 59,
    });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: options.id,
      privateKey: options.privateKey,
      cache,
      log: this.log,
      redisConfig: options.redisConfig,
      throttleOptions: options.throttleOptions,
    });
    const octokit = new Octokit();

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit,
      webhooks: {
        path: options.webhookPath,
        secret: options.secret,
      },
      id: options.id,
      privateKey: options.privateKey,
    };

    this.auth = auth.bind(null, this.state);

    this.webhooks = getWebhooks(this.state);

    this.on = (eventNameOrNames, callback) => {
      // when an app subscribes to an event using `app.on(event, callback)`, Probot sends a request to `GET /app` and
      // verifies if the app is subscribed to the event and logs a warning if it is not.
      //
      // This feature will be moved out of Probot core as it has side effects and does not work in a stateless environment.
      webhookEventCheck(this.state, eventNameOrNames);

      if (eventNameOrNames === "*") {
        // @ts-ignore this workaround is only to surpress a warning. The `.on()` method will be deprecated soon anyway.
        return this.webhooks.onAny(callback);
      }

      return this.webhooks.on(eventNameOrNames, callback);
    };

    this.server = createServer({
      webhook: (this.webhooks as any).middleware,
      logger: this.log,
    });

    const { version } = require("../package.json");
    this.version = version;

    // TODO: remove once Application class was removed
    this.internalRouter = express.Router();

    // TODO: Refactor tests so we we can remove these
    this.options = options;
  }

  /**
   * @deprecated `probot.webhook` is deprecated. Use `probot.webhooks` instead
   */
  public get webhook(): Webhooks {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.webhook" is deprecated. Use "probot.webhooks" instead`
      )
    );

    return this.webhooks;
  }

  public receive(event: WebhookEvent) {
    this.log.debug({ event }, "Webhook received");
    return this.webhooks.receive(event);
  }

  public load(appFn: string | ApplicationFunction | ApplicationFunction[]) {
    if (typeof appFn === "string") {
      appFn = resolveAppFunction(appFn) as ApplicationFunction;
    }

    const router = express.Router();

    // Connect the router from the app to the server
    this.server.use(router);

    // Initialize the ApplicationFunction
    load(this, router, appFn);

    return this;
  }

  public setup(appFns: Array<string | ApplicationFunction>) {
    // Log all unhandled rejections
    (process as NodeJS.EventEmitter).on(
      "unhandledRejection",
      getErrorHandler(this.log)
    );

    // Load the given appFns along with the default ones
    appFns.concat(defaultAppFns).forEach((appFn) => this.load(appFn));

    // Register error handler as the last middleware
    this.server.use(
      pinoHttp({
        logger: this.log,
      })
    );
  }

  public start() {
    this.log.info(
      `Running Probot v${this.version} (Node.js: ${process.version})`
    );
    const port = this.options.port || 3000;
    const { host } = this.options;
    const printableHost = host ?? "localhost";

    this.httpServer = this.server
      .listen(port, ...((host ? [host] : []) as any), () => {
        if (this.options.webhookProxy) {
          createWebhookProxy({
            logger: this.log,
            path: this.options.webhookPath,
            port: this.options.port,
            url: this.options.webhookProxy,
          });
        }
        this.log.info(`Listening on http://${printableHost}:${port}`);
      })
      .on("error", (error: NodeJS.ErrnoException) => {
        if (error.code === "EADDRINUSE") {
          this.log.error(
            `Port ${this.options.port} is already in use. You can define the PORT environment variable to use a different port.`
          );
        } else {
          this.log.error(error);
        }
        process.exit(1);
      });

    return this.httpServer;
  }

  public stop() {
    if (!this.httpServer) return;

    this.httpServer.close();
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = ({ app, getRouter }) => {
   *   // Get an express router to expose new HTTP endpoints
   *   const router = getRouter('/my-app');
   *
   *   // Use any middleware
   *   router.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   router.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes* @param path
   *
   * @deprecated "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"
   */
  route(path?: string) {
    this.log.warn(
      new Deprecation(
        `[probot] "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return getRouter(this.internalRouter, path);
  }
}

export const createProbot = (options: Options) => {
  options.log = options.log || getLog();
  options.log.warn(
    new Deprecation(
      `[probot] "createProbot(options)" is deprecated, use "new Probot(options)" instead`
    )
  );
  return new Probot(options);
};

export { Logger, Context, Application, ProbotOctokit };

/** NOTE: exported types might change at any point in time */
export { WebhookPayloadWithRepository };
