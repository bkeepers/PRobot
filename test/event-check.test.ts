import nock from 'nock'
import { Application } from '../src'
import eventCheck, { resetEventCheckCaches } from '../src/event-check'
import { newApp } from './apps/helper'

function mockAppMetaRequest (events: string[] = ['issues']) {
  return { events }
}

describe('event-check', () => {
  let app: Application
  let originalNodeEnv: string
  const unsubscribedEventName = 'an-unsubscribed-event.action'

  beforeAll(() => {
    originalNodeEnv = process.env.NODE_ENV || 'test'
  })

  beforeEach(() => {
    delete process.env.DISABLE_EVENT_CHECK
    nock.cleanAll()
    resetEventCheckCaches()
  })

  test('caches result of /app', async () => {
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest(['label', 'star']))
    app = newApp()
    const spyOnAuth = jest.spyOn(app, 'auth')
    const spyOnLogError = jest.spyOn(app.log, 'error')
    expect(await eventCheck(app, 'label.edited')).toStrictEqual(true)
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest(['team']))
    expect(await eventCheck(app, 'label.deleted')).toStrictEqual(true)
    expect(await eventCheck(app, 'team.created')).toStrictEqual(false)
    expect(spyOnAuth).toHaveBeenCalledTimes(1)
    expect(spyOnLogError).toMatchSnapshot()
  })

  test('succeeds when event name is wildcard character', async () => {
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest([]))
    app = newApp()
    expect(await eventCheck(app, '*')).toStrictEqual(true)
  })

  describe('warn user when', () => {
    test('listening to unsubscribed event', async () => {
      nock('https://api.github.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .get('/app').reply(200, mockAppMetaRequest())
      app = newApp()
      const spyOnLogError = jest.spyOn(app.log, 'error')
      expect(await eventCheck(app, unsubscribedEventName)).toStrictEqual(false)
      expect(spyOnLogError).toHaveBeenCalledTimes(1)
      expect(spyOnLogError).toMatchSnapshot()
    })

    test('unable to retrieve app metadata', async () => {
      nock('https://api.github.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .get('/app').reply(404)
      app = newApp()
      const spyOnLogWarn = jest.spyOn(app.log, 'warn')
      expect(await eventCheck(app, unsubscribedEventName)).toBeUndefined()
      expect(await eventCheck(app, `another-${unsubscribedEventName}`)).toBeUndefined()
      expect(spyOnLogWarn).toHaveBeenCalledTimes(1)
      expect(spyOnLogWarn).toMatchSnapshot()
    })
  })

  describe('can be disabled', () => {
    beforeEach(() => {
      delete process.env.DISABLE_EVENT_CHECK
      delete process.env.NODE_ENV
    })

    test('when DISABLE_EVENT_CHECK is true', async () => {
      process.env.DISABLE_EVENT_CHECK = 'true'
      app = newApp()
      const spyOnAuth = jest.spyOn(app, 'auth')
      expect(await eventCheck(app, 'issues.opened')).toBeUndefined()
      expect(spyOnAuth).toHaveBeenCalledTimes(0)
    })

    test('when NODE_ENV is production', async () => {
      process.env.NODE_ENV = 'production'
      app = newApp()
      const spyOnAuth = jest.spyOn(app, 'auth')
      expect(await eventCheck(app, 'issues.opened')).toBeUndefined()
      expect(spyOnAuth).toHaveBeenCalledTimes(0)
    })
  })

  afterEach(() => {
    process.env.NODE_ENV = originalNodeEnv
  })
})
