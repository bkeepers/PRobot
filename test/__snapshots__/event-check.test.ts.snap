// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event-check caches result of /app 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"team.created\\", but your GitHub App is not subscribed to the \\"team\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`event-check warn user when listening to unsubscribed event 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"an-unsubscribed-event.action\\", but your GitHub App is not subscribed to the \\"an-unsubscribed-event\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`event-check warn user when unable to retrieve app metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Probot is unable to retrieve app information from GitHub for event subscription verification.

If this error persists, feel free to raise an issue at:
  - https://github.com/probot/probot/issues",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
